desc/objectives/utils.py

```diff
@@ -41,40 +54,50 @@ def factorize_linear_constraints(constraints, objective_args):  # noqa: C901
         and recovering x from y.
 
     """
+    for con in constraints:
+        for thing in con.things:
+            if thing not in objective.things:
+                warnings.warn(
+                    f"Optimizable object {thing} is constrained"
+                    + " but not included in Objective"
+                )
     # set state vector
-    args = np.concatenate([obj.args for obj in constraints])
-    args = np.concatenate((args, objective_args))
-    # this is all args used by both constraints and objective
-    args = [arg for arg in arg_order if arg in args]
-    dimensions = constraints[0].dimensions
-    dim_x = 0
-    x_idx = {}
-    for arg in args:
-        x_idx[arg] = np.arange(dim_x, dim_x + dimensions[arg])
-        dim_x += dimensions[arg]
-
+    xp = jnp.zeros(objective.dim_x)  # particular solution to Ax=b
     A = []
     b = []
-    xp = jnp.zeros(dim_x)  # particular solution to Ax=b
+
+    from desc.optimize import ProximalProjection
+
+    prox_flag = isinstance(objective, ProximalProjection)
 
     # linear constraint matrices for each objective
-    for obj_ind, obj in enumerate(constraints):
-        if obj.bounds is not None:
-            raise ValueError("Linear constraints must use target instead of bounds.")
-        A_ = {
-            arg: obj.derivatives["jac_scaled"][arg](
-                *[jnp.zeros(obj.dimensions[arg]) for arg in obj.args]
+    for con in constraints:
+        if con.bounds is not None:
+            raise ValueError(
+                f"Linear constraint {con} must use target instead of bounds."
             )
-            for arg in args
-        }
+        A_per_thing = []
+        xz = _tree_zeros_like([t.params_dict for t in con.things])
+        # computing A matrix for each constraint for each thing in the optimization
+        for thing in objective.things:
+            if thing in con.things:
+                # for now we implicitly assume that each linear constraint is bound to
```

Comment:
what would be an example of a generalization? something like constraining the R0 of every coil in a coilset or something?

Reply:
yeah, or the parameter sharing thing we discussed. Still need to think about how that will work but can probably punt to the next PR